publish.go

```diff
@@ -0,0 +1,357 @@
+	if err := updateListing(configs, release); err != nil {
+		return fmt.Errorf("failed to update listing, reason: %v", err)
+	}
+	return nil
+}
+
+// trackNamesToUpdate returns the tracks that should be updated.
+func trackNamesToUpdate(track string, tracks []*androidpublisher.Track) []string {
+	var possibleTrackNamesToUpdate []string
+	switch track {
+	case betaTrackName:
+		possibleTrackNamesToUpdate = []string{alphaTrackName}
+	case rolloutTrackName, productionTrackName:
+		possibleTrackNamesToUpdate = []string{alphaTrackName, betaTrackName}
+	}
+
+	var trackNamesToUpdate []string
+	for _, track := range tracks {
+		for _, trackNameToUpdate := range possibleTrackNamesToUpdate {
+			if trackNameToUpdate == track.Track {
+				trackNamesToUpdate = append(trackNamesToUpdate, trackNameToUpdate)
+			}
+		}
+	}
+
+	log.Printf(" possible tracks to update: %v", trackNamesToUpdate)
+	return trackNamesToUpdate
+}
+
+// unTrackFromTracks untracks the lower level versions.
+func unTrackFromTracks(trackNamesToUpdate []string, versionCodes googleapi.Int64s, service *androidpublisher.Service, packageName string, appEditID string) error {
+	tracksService := androidpublisher.NewEditsTracksService(service)
+	anyTrackUpdated := false
+	for _, trackName := range trackNamesToUpdate {
+		tracksGetCall := tracksService.Get(packageName, appEditID, trackName)
+		track, err := tracksGetCall.Do()
+		if err != nil {
+			return fmt.Errorf("failed to get track (%s), error: %s", trackName, err)
+		}
+		for _, release := range track.Releases {
+			if removeBlockingVersionFromRelease(release, versionCodes) {
+				anyTrackUpdated = true
+
+				release.VersionCodes = []int64{}
+				release.NullFields = []string{"VersionCodes"}
+				release.ForceSendFields = []string{"VersionCodes"}
+			}
+		}
+
+		if anyTrackUpdated {
+			log.Donef("Desired versions deactivated")
+		} else {
+			log.Donef("No blocking apk version found")
+		}
+		tracksUpdateCall := tracksService.Patch(packageName, appEditID, trackName, track)
+		if _, err := tracksUpdateCall.Do(); err != nil && err != io.EOF {
+			return fmt.Errorf("failed to update track (%s), error: %s", trackName, err)
+		}
+	}
+	return nil
+}
+
+// removeBlockingVersionFromRelease removes blocking version from a given version, which would shadow the given version.
+func removeBlockingVersionFromRelease(release *androidpublisher.TrackRelease, newVersionCodes googleapi.Int64s) bool {
+	log.Printf("Checking app versions on release: %s", release.Name)
+	log.Infof("Current version codes: %v", release.VersionCodes)
+	log.Infof("New version codes: '%v'", newVersionCodes)
+
+	if len(release.VersionCodes) != len(newVersionCodes) {
+		log.Warnf("Mismatching app count, removing (%v) versions from release: %s", release.VersionCodes, release.Name)
+		return true
+	}
+
+	log.Debugf("The number of App version codes (current and new) are equal")
+	return sortAndFilterVersionCodes(release.VersionCodes, newVersionCodes)
```

Comment:
It seems that sortAndFilterVersionCodes is supposed to mutatue the versionCodes but from the function definition it seems that does not to that.
Can you add unit tests for removeBlockingVersionFromRelease?
Make sure to update functon names/docs for the changes in https://github.com/bitrise-steplib/steps-google-play-deploy/pull/63/commits/900ac4d5700baa94e50d8eca7ca27342f7d00d9b.

Reply:
Updated and added test cases.